### ISSUE TITLE
FIX: PackageKit.sh does not check if packagekitd is running at all

### ISSUE THREAD

**Issue by Cyborgscode** [PackageKit/PackageKit#521]
Fix for: 

$ 91.216.198.43
bash: 91.216.198.43: command not found...
Failed to search for file: GDBus.Error:org.freedesktop.DBus.Error.NameHasNoOwner: Could not activate remote peer.

by patching PackageKit.sh :

command_not_found_handle () {
	local runcnf=1
	local retval=127

	# only search for the command if we're interactive
	[[ $- == *"i"* ]] || runcnf=0

	# don't run if DBus isn't running
	[[ ! -S /run/dbus/system_bus_socket ]] && runcnf=0

	# don't run if packagekitd doesn't exist in the _system_ root
	[[ ! -x '/usr/libexec/packagekitd' ]] && runcnf=0

 	**# only run cnf, if packagekitd is actually running to avoid errormessages!
 
 	RC=$(pgrep -c packagekitd)
 
 	[[ "$RC" == "0" ]] && runcnf=0**

NOTE:

No, i don't clone an entire projekt, just to supply 3 lines of bash code.

**Comment by hughsie** [comment]
> No, i don't clone an entire projekt, just to supply 3 lines of bash code

That's not the correct fix, CNF needs to be able to autostart packagekitd when it's not running. You need to handle `Could not activate remote peer` properly and mask the error.

**Comment by ximion** [comment]
Shouldn't D-Bus activation autostart the daemon as soon as the pk-command-no-found helper queries it?

**Comment by Cyborgscode** [comment]
let me explain why it's not working :

packagekitd was masked out in systemd, so there is no way that the cnf script or any other can start pckagekitd.

And here is the actual use case:

The Fedora PC in question uses Cinnamon as main Desktop Environment  and Gnome as a reserve DE for RDP.
Gnome Software isn't used to install software, so packagekit does not need to work. It also "wasted" ressources on startup
and in mid work, so it got decided to disable it, but because of how it's integrated, restarted on calls like the cnf a.s.o. it needed to be masked.

@Richard: you see, it's wanted that the PackageKit.sh script can't start it. 

I admit, that my fix attempt in PackageKit.sh, check if the daemon is running, contradicts the possibility of starting the service later in the event chain. But, if the start attempt for a  not running service is done in a different handler, i suggest to keep the fix in PackageKit.sh, because that start attempt would fail for our wanted scenario, but would succeed for any other scenario, where packagekit it startable. 

A win-win: We get rid of the annoying error message, and keep packagekit as dependency for gnome intact, and everyone else is still happy with lazy start of packagekitd. 

I really don't want to deinstall cnf , but if thats the best option in your opinion, than it's also fine, but i have the impression, other event will write error message for the masked out PK too, so checking there, if it's really possible to access PK, would be a smart move too.

**Comment by hughsie** [comment]
> and everyone else is still happy with lazy start of packagekitd

I don't agree. It's either deterministic, or not. Given packagekitd often auto-quits on idle this would be very confusing.

> I really don't want to deinstall cnf , but if thats the best option in your opinion, than it's also fine,

What's the problem with just removing the CNF subpackage?

**Comment by Cyborgscode** [comment]
Multiuser System.. not everyone may wanne miss the functionality. You never know when someone needs it. 

If PackageKit.sh is the wrong place  and "/usr/libexec/pk-command-not-found"  wants to start packagekitd,
why isn't there a check, if the start actually worked at all before accessing it? 

To solve my issue, i will now remove cnf from the system, but i still think, that there is a logic error with this cornercase. One can't assume, that just because a daemon is installed, it is also runable.

**Comment by hughsie** [comment]
> not everyone may wanne miss the functionality. You never know when someone needs it

Isn't masking the service effective for all users?

**Comment by Cyborgscode** [comment]
it is. Ofcourse, someone needs to reenable it, before a normal user can make use of it.
